File: src/first_misread/analyzer.py
"""Content analysis — structural metadata extraction."""

from __future__ import annotations

import re

from first_misread.models import ContentMetadata

WORDS_PER_MINUTE = 200


def analyze_content(text: str) -> ContentMetadata:
    """Analyze text and return structural metadata."""
    words = text.split()
    word_count = len(words)
    read_time = word_count / WORDS_PER_MINUTE

    paragraphs = [p.strip() for p in text.split("\n\n") if p.strip()]
    paragraph_count = len(paragraphs)

    headings = re.findall(r"^#{1,6}\s+.+$", text, re.MULTILINE)
    heading_count = len(headings)

    sentences = re.split(r"[.!?]+\s+", text)
    sentences = [s for s in sentences if s.strip()]
    sentence_count = len(sentences)

    avg_sentence_length = (
        sum(len(s.split()) for s in sentences) / sentence_count
        if sentence_count
        else 0
    )

    has_lists = bool(re.search(r"^[\s]*[-*]\s+", text, re.MULTILINE))
    has_links = bool(re.search(r"\[.+?\]\(.+?\)", text))

    return ContentMetadata(
        word_count=word_count,
        estimated_read_time_minutes=round(read_time, 1),
        paragraph_count=paragraph_count,
        heading_count=heading_count,
        has_lists=has_lists,
        has_links=has_links,
        sentence_count=sentence_count,
        avg_sentence_length=round(avg_sentence_length, 1),
    )

File: extension/src/core/analyzer.js
const WORDS_PER_MINUTE = 200;

export function analyzeContent(text) {
  const words = text.split(/\s+/).filter(Boolean);
  const wordCount = words.length;
  const readTime = Math.round((wordCount / WORDS_PER_MINUTE) * 10) / 10;

  const paragraphs = text.split(/\n\n/).filter(p => p.trim());
  const headings = text.match(/^#{1,6}\s+.+$/gm) || [];

  const sentences = text.split(/[.!?]+\s+/).filter(s => s.trim());
  const sentenceCount = sentences.length;
  const avgSentenceLength = sentenceCount
    ? Math.round(sentences.reduce((sum, s) => sum + s.split(/\s+/).filter(Boolean).length, 0) / sentenceCount)
    : 0;

  const hasLists = /^[\s]*[-*]\s+/m.test(text);
  const hasLinks = /\[.+?\]\(.+?\)/.test(text);

  return {
    wordCount,
    estimatedReadTimeMinutes: readTime,
    paragraphCount: paragraphs.length,
    headingCount: headings.length,
    hasLists,
    hasLinks,
    sentenceCount,
    avgSentenceLength,
  };
}
